File: 3310.remove-methods-from-project.py
#
# @lc app=leetcode id=3310 lang=python3
#
# [3310] Remove Methods From Project
#

# @lc code=start
class Solution:
    def remainingMethods(self, n: int, k: int, invocations: List[List[int]]) -> List[int]:
        from collections import defaultdict, deque
        # Build adjacency list for graph representation of method invocations
        adj_list = defaultdict(list)
        reverse_adj_list = defaultdict(list)
        for ai, bi in invocations:
            adj_list[ai].append(bi)
            reverse_adj_list[bi].append(ai)
        
        # Collect all reachable nodes from k (suspicious nodes) using DFS
        def collect_suspicious(start):
            stack = [start]
            visited = set()
            while stack:
                node = stack.pop()
                if node not in visited:
                    visited.add(node)
                    stack.extend(adj_list[node])
            return visited
        
        suspicious_methods = collect_suspicious(k)
        # Check for external invocation to suspicious nodes
        for sm in suspicious_methods:
            if any(invoker not in suspicious_methods for invoker in reverse_adj_list[sm]):
                return list(range(n))  # Return all methods because we can't remove them cleanly.
        
        # Return non-suspicious methods only if possible to remove all suspicious ones.
        return [i for i in range(n) if i not in suspicious_methods]
# @lc code=end

File: 3310.remove-methods-from-project.go
#
# @lc app=leetcode id=3310 lang=golang
#
# [3310] Remove Methods From Project
#

# @lc code=start
func remainingMethods(n int, k int, invocations [][]int) []int {
    // Create adjacency lists for the graph representation
    adjList := make([][]int, n)
    reverseAdjList := make([][]int, n)
    for _, invocation := range invocations {
        adjList[invocation[0]] = append(adjList[invocation[0]], invocation[1])
        reverseAdjList[invocation[1]] = append(reverseAdjList[invocation[1]], invocation[0])
    }
    
    // Find all suspicious methods using DFS starting from method k
    suspicious := make([]bool, n)
    var dfs func(node int)
    dfs = func(node int) {
        if suspicious[node] {
            return
        }
        suspicious[node] = true
        for _, nei := range adjList[node] {
            dfs(nei)
        }
    }
    dfs(k)
    
    // Check for any incoming edges to the suspicious set from outside
    canRemoveSuspicious := true
    for i := 0; i < n; i++ {
        if !suspicious[i] { // Only consider non-suspicious methods
            for _, nei := range adjList[i] {
                if suspicious[nei] { // Found an incoming edge to a suspicious method
                    canRemoveSuspicious = false
                    break
                }
            }
        }
        if !canRemoveSuspicious {
            break
        }
    }
    
    // Determine result based on whether we can safely remove suspicious methods or not
    result := []int{}
    if canRemoveSuspicious {
        for i := 0; i < n; i++ {
            if !suspicious[i] {
                result = append(result, i)
            }
        }
    } else {
        for i := 0; i < n; i++ {
            result = append(result, i)
        }
    }
    return result
}
# @lc code=end